### ISSUE TITLE
[Bug] mclmc_adaptation.handle_nans never observes NaNs during tuning. step_size_max does not shrink as a result.

### ISSUE THREAD

**Issue by ssage0520** [blackjax-devs/blackjax#969]
### Checklist

- [x] I have searched existing issues and this is not a duplicate.
- [x] I can reproduce this with the latest version of BlackJAX.

### Describe the bug

mclmc_adaptation.handle_nans is supposed to detect divergent (NaN) steps during mclmc_find_L_and_step_size tuning, revert to the last good state, and shrink step_size_max so the next proposed step size gets clipped down. 

However, the mclmc kernel has internal NaN handling with its handle_nans function. Any divergent step is already reverted to the previous state before being returned to the caller. As a result, mclmc_adaptation.py's handle_nans never observes a non-finite next_state during tuning. The nonans check is always True, and the step_size_max-shrinking branch can no longer execute in practice.

This also affects the predictor function in make_L_step_size_adaptation. The energy_change gets zeroed, pushing up the step size. The intent is that step_size_max will counteract that increase, but in practice step_size_max was not properly updated.

### Minimal reproducible example

```python
import jax
import jax.numpy as jnp
from blackjax.mcmc.mclmc import init, build_kernel
from blackjax.adaptation.mclmc_adaptation import (
    handle_nans as adaptation_handle_nans,
    mclmc_find_L_and_step_size,
    MCLMCAdaptationState,
)

dim = 2
bound = 5.0

def logdensity_fn(x):
    # Finite only on (-bound, bound)
    return -0.5 * jnp.sum(x**2) + jnp.sum(jnp.log(bound - jnp.abs(x)))

rng_key = jax.random.key(0)
init_key, step_key = jax.random.split(rng_key)
state = init(jnp.zeros(dim), logdensity_fn, init_key)
kernel = build_kernel()

initial_step_size = 100.0  # guarantees the first step diverges

# 1. The kernel detects the divergence
next_state, info = kernel(
    rng_key=step_key, state=state, logdensity_fn=logdensity_fn,
    inverse_mass_matrix=1.0, L=1.0, step_size=initial_step_size,
)
print("kernel info.nonans:", info.nonans)                    # False
print("kernel next_state.position:", next_state.position)    # initial position [0. 0.]

# 2. By the time mclmc_adaptation.handle_nans sees it, there's
#    nothing left to detect.
success, _, step_size_max, _ = adaptation_handle_nans(
    previous_state=state, next_state=next_state, step_size=initial_step_size,
    step_size_max=jnp.inf, kinetic_change=info.energy_change, key=step_key,
)
print("adaptation handle_nans success:", success)            # True (should be False)
print("adaptation step_size_max:", step_size_max)            # unshrunk

# 3. mclmc_find_L_and_step_size increases the step size instead of shrinking it.
initial_params = MCLMCAdaptationState(
    L=1.0, step_size=initial_step_size, inverse_mass_matrix=jnp.ones(dim)
)
_, params, _ = mclmc_find_L_and_step_size(
    mclmc_kernel=kernel,
    num_steps=1,
    state=state,
    rng_key=step_key,
    logdensity_fn=logdensity_fn,
    params=initial_params,
    frac_tune1=1.0, frac_tune2=0.0, frac_tune3=0.0,
    diagonal_preconditioning=False,
)
print("step_size before:", initial_step_size)
print("step_size after one divergent tuning step:", params.step_size)
```

### Expected behaviour

When the underlying mclmc kernel encounters a divergence during tuning, mclmc_adaptation's own NaN handling should be able to detect it and shrink step_size_max accordingly — per its docstring, "if there are nans, let's reduce the stepsize." Concretely, given a kernel call whose info.nonans is False (a real divergence occurred), feeding that step's output into mclmc_adaptation.handle_nans should report success=False and reduce step_size_max below the step size that caused the divergence (e.g. step_size * 0.8). A subsequent call to mclmc_find_L_and_step_size starting from a step size that just caused a divergence should not increase that step size.

### Actual behaviour / error message

```shell
mclmc_adaptation.handle_nans reports success=True even when the kernel's own info.nonans was False for that same step, because by the time handle_nans inspects next_state, the kernel has already reverted it to a clean, finite state. step_size_max is therefore never shrunk (it comes back as nan_to_num(inf) ≈ 3.4e38, i.e. effectively unconstrained). As a direct consequence, mclmc_find_L_and_step_size, run for a single tuning step starting from step_size=100.0 (which diverges), returns step_size≈2154.4 — roughly a 20x increase immediately following a divergence.
```

### Version and environment info

```python
BlackJAX 1.6.dev98+gf73e12956
JAX 0.9.1
jaxlib 0.9.1
Python 3.12.13 | packaged by conda-forge | (main, Mar  5 2026, 16:50:00) [GCC 14.3.0]
Devices: [CpuDevice(id=0)]
```

### Additional context

Also reproduces on the current PyPI release, blackjax==1.5 (confirmed in a clean venv). The kernel-construction API differs slightly between 1.5 and main, so the MRE above targets main. I am happy to provide a 1.5-adapted version if useful.

**Comment by junpenglao** [comment]
Thank you for the excellent report — the MRE reproduces exactly at HEAD, and your mechanism is confirmed. We ran a two-arm investigation and can add three findings:

1. **Severity**: it's worse than a one-time ~20× bump — once a tuning step goes non-finite, the step size grows monotonically with no escape (measured e.g. 100 → 2154 → … → 6810 as tuning length grows; ~75% of seeds run away on a bounded target at `initial_step_size=100`). The `step_size_max` shrink you identified as dead is the only counterweight, so its death makes the loop unrecoverable.

2. **Your case is one of two divergence signatures, and the kernel misses the other one entirely.** Your MRE (McLachlan, L=1.0) produces position/momentum-NaN → the kernel reverts and reports `nonans=False` (case-1). But the kernel's finiteness check covers **position and momentum only — not `logdensity`** — so a step landing at a *finite* point outside support (NaN logdensity, finite gradient) is neither reverted nor flagged (case-2). Case-2 dominates under `velocity_verlet` (24/24 seeds at step sizes 8–20 on the same bounded target) and fires mid-trajectory even on McLachlan. Consequence: fixing this by consuming `info.nonans` alone is insufficient.

3. **The case-2 root cause also corrupts the plain sampling path, independent of adaptation**: at a too-large fixed step size, the kernel caches the NaN `logdensity` in the returned state and propagates it (9/15 steps at `velocity_verlet` step_size=8) while reporting `nonans=True` throughout — silent corruption with no divergence flag.

**Fix in progress** (PR shortly): (i) kernel-side root fix — extend the kernel's `handle_nans` finiteness test to include `logdensity`, making `info.nonans` truthful for both signatures and fixing the sampling-path propagation; (ii) the adaptation's `handle_nans` consumes the kernel's flag (plus an `isfinite(energy_change)` guard) instead of re-deriving from the already-sanitized state — restoring the `step_size_max` shrink you diagnosed; (iii) the same dead-signal pattern in the LAPS burn-in gets the same treatment. The regression traces to #806, which introduced the kernel-internal revert and the `nonans` info field together without wiring the adaptation to consume it.

Bitwise non-regression on NaN-free targets (the common case) is part of the test plan — on well-behaved targets the fix is a provable no-op.

— 🤖 Blackjax-devs AI TL
